### PR TITLE
Make `TyKind` `Copy` and change `ty.kind()` to return `TyKind`

### PR DESCRIPTION
**Pull request by LeSeulArtichaut**
Implements T-compiler MCP rust-lang/compiler-team#363.
This should be perf-tested.

r? @lcnr cc @nikomatsakis

### CODE REVIEW

**Comment by lcnr** [comment]
:heart: 

@bors try @rust-timer perf

File: compiler/rustc_middle/src/ty/codec.rs [hunk 1]
```diff
@@ -39,7 +39,7 @@ impl<'tcx, E: TyEncoder<'tcx>> EncodableWithShorthand<'tcx, E> for Ty<'tcx> {
 
     #[inline]
     fn variant(&self) -> &Self::Variant {
-        self.kind()
+        &self.kind
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/codec.rs:42]: hmm, can you add a fixme here, I would like to potentially solve this differently.

Might just use
```suggestion
        // FIXME(@lcnr): Look into instead returning `Self::Variant` directly here
        &self.kind
```

File: compiler/rustc_middle/src/ty/error.rs [hunk 2]
```diff
@@ -353,7 +353,7 @@ impl<'tcx> TyCtxt<'tcx> {
                         );
                     }
                 }
-                match (values.expected.kind(), values.found.kind()) {
+                match (&values.expected.kind(), &values.found.kind()) {
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/error.rs:356]: why are you matching on `&values.expected.kind()` here?

File: compiler/rustc_middle/src/ty/error.rs [hunk 3]
```diff
@@ -575,7 +575,7 @@ impl<T> Trait<T> for X {
                 // This will also work for `impl Trait`.
                 let def_id = if let ty::Param(param_ty) = proj_ty.self_ty().kind() {
                     let generics = self.generics_of(body_owner_def_id);
-                    generics.type_param(param_ty, self).def_id
+                    generics.type_param(&param_ty, self).def_id
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/error.rs:578]: Probably want to instead take `ParamTy` by value in `type_param`, but using a reference is fine for now here

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 4]
```diff
@@ -1118,7 +1118,7 @@ pub trait PrettyPrinter<'tcx>:
 
         let u8_type = self.tcx().types.u8;
 
-        match (ct, ty.kind()) {
+        match (ct, &ty.kind()) {
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/print/pretty.rs:1121]: `&ty.kind()`?

File: compiler/rustc_middle/src/ty/relate.rs [hunk 5]
```diff
@@ -325,7 +325,7 @@ pub fn super_relate_tys<R: TypeRelation<'tcx>>(
 ) -> RelateResult<'tcx, Ty<'tcx>> {
     let tcx = relation.tcx();
     debug!("super_relate_tys: a={:?} b={:?}", a, b);
-    match (a.kind(), b.kind()) {
+    match (&a.kind(), &b.kind()) {
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/relate.rs:328]: ref kind?

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 6]
```diff
@@ -347,7 +347,7 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                         "`Self`",
                         err,
                         None,
-                        projection,
+                        projection.as_ref(),
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:350]: followup: probably take `Option<ProjectionTy>` in suggest_restriction

File: compiler/rustc_typeck/src/check/op.rs [hunk 7]
```diff
@@ -569,9 +569,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         };
 
         match (lhs_ty.kind(), rhs_ty.kind()) {
-            (&Ref(_, l_ty, _), &Ref(_, r_ty, _)) // &str or &String + &str, &String or &&str
-                if (*l_ty.kind() == Str || is_std_string(l_ty)) && (
-                        *r_ty.kind() == Str || is_std_string(r_ty) ||
+            (Ref(_, l_ty, _), Ref(_, r_ty, _)) // &str or &String + &str, &String or &&str
+                if (l_ty.kind() == Str || is_std_string(l_ty)) && (
+                        r_ty.kind() == Str || is_std_string(r_ty) ||
                         &format!("{:?}", rhs_ty) == "&&str"
```
Comment (lcnr) [inline compiler/rustc_typeck/src/check/op.rs:575]: unrelated, but we really shouldn't compare on `format!("{:?}", rhs_ty)` here

**Review by lcnr — APPROVED** [review]
some small nits, otherwise this is a lot cleaner than before. Thanks!

**Comment by lcnr** [comment]
used the wrong rust-timer command

@rust-timer build

**Comment by LeSeulArtichaut** [comment]
Also cc @eddyb in case you're interested in the perf results

**Comment by lcnr** [comment]
not sure about perf, cycles looks quite positive, wall-time is probably about neutral all considered and instructions got worse.

I would have expected that we mostly just instantly match on `ty.kind()` which requires loading it from memory anyways, so I don't fully get why this increases the instruction count here.

**Comment by lcnr** [comment]
I think using `ref` patterns here ends up causing worse codegen when returning `TyKind` in kind.

So we probably get better perf - and cleaner code - if we eliminate these.

experimented in https://godbolt.org/z/o3bsM3

File: compiler/rustc_middle/src/ty/relate.rs [hunk 8]
```diff
@@ -325,8 +325,8 @@ pub fn super_relate_tys<R: TypeRelation<'tcx>>(
 ) -> RelateResult<'tcx, Ty<'tcx>> {
     let tcx = relation.tcx();
     debug!("super_relate_tys: a={:?} b={:?}", a, b);
-    match (&a.kind(), &b.kind()) {
-        (&ty::Infer(_), _) | (_, &ty::Infer(_)) => {
+    match (a.kind(), b.kind()) {
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/relate.rs:328]: I hope this was the main reason for the regression. Relate is quite hot and `&a.kind()` has quite bad codegen

File: compiler/rustc_middle/src/ty/relate.rs [hunk 9]
```diff
@@ -335,39 +335,39 @@ pub fn super_relate_tys<R: TypeRelation<'tcx>>(
...
+        | (ty::Uint(_), _)
+        | (ty::Float(_), _)
+        | (ty::Str, _)
             if a == b =>
         {
             Ok(a)
         }
 
-        (&ty::Param(ref a_p), &ty::Param(ref b_p)) if a_p.index == b_p.index => Ok(a),
+        (ty::Param(ref a_p), ty::Param(ref b_p)) if a_p.index == b_p.index => Ok(a),
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/relate.rs:352]:
```suggestion
        (ty::Param(a_p), ty::Param(b_p)) if a_p.index == b_p.index => Ok(a),
```

**Comment by lcnr** [comment]
let's try again

@bors try @rust-timer queue

**Comment by lcnr** [comment]
See https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/Make.20TyKind.20Copy.20and.20change.20ty.2Ekind%28%29.20to.20.E2.80.A6.20compiler-team.23363/near/212460531 for some discussion about this.

A huge thanks for @LeSeulArtichaut for implementing this MCP!

**Pull request closed without merging**